[NF-8840] Travis build hook for Rails 4 compat

kstevens715 opened the pull request:
See lib/ci/README.md for more details. This creates a script that
watches for Travis builds to complete. When one completes, if it's a PR
it will count the number of failures, errors and passing tests for each
job that allows failures (Rails 4 jobs). It then compares those against
the equivalent jobs from the last build to finish against the master
branch. Then it takes the results and posts a comment to the GitHub pull
request with the results. If the pull request makes any sub-build worse,
it labels the PR as Rails 4 Rejected. If it improves the situation or
leaves it the same it will label it Rails 4 Accepted. If the master
build and feature branch build do not have the same number of
sub-builds, it labels it as needing to be rebased against master.

Deploying this to GCE or running it locally requires you to set the
GITHUB_TOKEN environment variable with a personal access token that you
must setup. See:
https://help.github.com/articles/creating-an-access-token-for-command-line-use/
## Unexpanded

![selection_003](https://cloud.githubusercontent.com/assets/1191305/18219712/d1a60b76-7139-11e6-8627-83b2dc022e33.png)
## Expanded

![selection_004](https://cloud.githubusercontent.com/assets/1191305/18219716/d523345e-7139-11e6-8230-43db03a44418.png)

kstevens715 committed:
[NF-8840] Travis build hook for Rails 4 compat

See lib/ci/README.md for more details. This creates a script that
watches for Travis builds to complete. When one completes, if it's a PR
it will count the number of failures, errors and passing tests for each
job that allows failures (Rails 4 jobs). It then compares those against
the equivalent jobs from the last build to finish against the master
branch. Then it takes the results and posts a comment to the GitHub pull
request with the results. If the pull request makes any sub-build worse,
it labels the PR as Rails 4 Rejected. If it improves the situation or
leaves it the same it will label it Rails 4 Accepted. If the master
build and feature branch build do not have the same number of
sub-builds, it labels it as needing to be rebased against master.

Deploying this to GCE or running it locally requires you to set the
GITHUB_TOKEN environment variable with a personal access token that you
must setup. See:
https://help.github.com/articles/creating-an-access-token-for-command-line-use/

benjaminoakes commented on lib/ci/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+RUN gem install bundler
+
+
+COPY ./ ./
```
I'm not sure if this is how we want to approach this...  It's also a little confusing copying `./` to `./`.

kstevens715 replied:
I might be wrong, but my thought was if we want to deploy this to a server, rather than running it locally we would need to copy the source files (rather than use the `-v` option). Because that's the only way to include it in the image itself, right? Or maybe you mean something different...

benjaminoakes replied:
I think you're correct, and I'm not currently sure of the "right way" to do this.

benjaminoakes commented on lib/ci/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+RUN bundle install
+
+RUN echo 'export PATH="/opt/bin:$PATH"' > /etc/profile.d/upgrade_toolbelt.sh
```
I don't think you're doing anything with `/opt`, are you?

kstevens715 replied:
Yeah, I can probably remove this. I started out adding one line at a time, but hit some snags and ended up copying some things verbatim from your Dockerfile.

kstevens715 commented on lib/ci/test/github_proxy_test.rb:
```diff
@@ -0,0 +1,56 @@
+require "minitest/autorun"
```
Maybe I should move the "test" directory to the root level and add this project to Travis. I added tests just to help me out during development, but if we don't run them they'll end up becoming broken and useless.

kstevens715 committed:
[NF-8840] Add CI to project

kstevens715 committed:
[NF-8840] Cache bundle install

By copying the Gemfile over first, Docker can cache the bundle step
making it much faster to build a new image. Thanks Pierre!

kstevens715 committed:
[NF-8840] Don't analyze builds that aren't PR's

kstevens715 committed:
[NF-8840] Extract deprecation logic to class

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,142 @@
+
+    def run(build)
+      unless build.pull_request?
+        puts "Build #{} is not a pull request. Skipping."
```
Missing something in those `#{}` brackets?

kstevens715 committed:
[NF-8840] Add deprecation warnings

kstevens715 committed:
[NF-8840] Add HTML expanders and improve format

This adds HTML5 expanders so the comment doesn't take up the full pull
request screen.

kstevens715 committed:
[NF-8840] Add missing variable for interpolation

devdude commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+      end.compact
+    end
+
+    def report_results(build, results, master_results)
```
Most likely this is a WIP since this is a super long method.

benjaminoakes replied:
Yeah, feels like we should extract some smaller pieces maybe?

devdude commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+          github.add_labels_to_an_issue([RAILS_4_WARNING])
+        end
+
+        comment << "<details>"
```
We could use erb templates to decouple the logic to render the output.

benjaminoakes replied:
For what it's worth, I don't think this expander works as intended in Firefox (and other non-WebKit-lineage browsers).  That's not something we can do much about, just a "known issue".

benjaminoakes replied:
Could use Mustache as well... (Possibly easier to use in this case.)  But regardless, there are other things that seem higher priority from my standpoint.

devdude commented on lib/ci/lib/rails_upgrade_analyzer/result_reporter.rb:
```diff
@@ -0,0 +1,112 @@
+    end
+
+    def build_deprecation_report
+      builder = Builder::XmlMarkup.new(indent: 2)
```
ERB templates for rendering the output.

devdude commented on lib/ci/lib/rails_upgrade_analyzer/result_reporter.rb:
```diff
@@ -0,0 +1,112 @@
+    end
+
+    def build_report
+      builder = Builder::XmlMarkup.new(indent: 2)
```
Sounding like a broken record now. ERB templates...

devdude commented:
Setting a good precedent for useful scripts with this one.

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+module RailsUpgradeAnalyzer
+  class Listener
+
+    REPO_NAME = "Liaison-Intl/WebAdMIT"
```
This shouldn't be in a public repo.

benjaminoakes replied:
Can we have this come from an env variable?

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+
+      login
+
+      Travis::Pro.listen(webadmit) do |stream|
```
`webadmit` isn't a generic name; this is a public repo.

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+    def run_one(build_number)
+      login
+
+      build = webadmit.build(build_number)
```
`webadmit` isn't a generic name; this is a public repo.

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+    end
+
+    def last_complete_master_build
+      webadmit.builds.detect do |build|
```
`webadmit` isn't a generic name; this is a public repo.

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+
+    def analyze_job(job, name)
+      puts "Analyzing job: #{job.number}"
+      link = "<a href='https://travis-ci.com/Liaison-Intl/WebAdMIT/jobs/#{job.id})'>#{name}</a>"
```
`Liaison-Intl/WebAdMIT` isn't a generic link; this is a public repo.

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+    end
+
+    def webadmit
+      @webadmit ||= Travis::Pro::Repository.find(REPO_NAME)
```
`webadmit` isn't a generic name; this is a public repo.

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+  end
+end
+
+#RailsUpgradeAnalyzer::Listener.new.run_one("16506")
```
This comment probably should be removed before merging.

benjaminoakes commented on Gemfile.lock:
```diff
@@ -0,0 +1,68 @@
+  travis
+
+RUBY VERSION
+   ruby 2.1.4p265
```
Doesn't match the `2.3.1` we have elsewhere...

benjaminoakes commented on bin/utb-deprecations-categorize:
```diff
@@ -1,36 +1,21 @@
 
 puts
-puts "Total: #{total} deprecations"
+puts "Total: #{summary.deprecation_count} deprecations"
```
Thanks for refactoring this!  🎉

benjaminoakes commented on lib/ci/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+RUN pkg-apk build-base=0.4-r1
+RUN pkg-apk libffi-dev=3.2.1-r2
+RUN pkg-apk ruby-irb=2.3.1-r0
+RUN pkg-gem rdoc
```
Why do we need `rdoc`?

benjaminoakes commented on lib/ci/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+RUN pkg-apk libffi-dev=3.2.1-r2
+RUN pkg-apk ruby-irb=2.3.1-r0
+RUN pkg-gem rdoc
+RUN pkg-gem io-console
```
There's an `apk` package for this, since it's normally part of Ruby.

benjaminoakes replied:
Also, we may want to pin these versions...

benjaminoakes commented on lib/ci/Dockerfile:
```diff
@@ -0,0 +1,28 @@
+COPY Gemfile.lock Gemfile.lock
+RUN bundle install
+
+COPY . .
```
This still seems like a hard thing to use in development.  I'd like to get @pfcarrier's thoughts on this, especially considering some other work he's done recently.

pfcarrier replied:
@benjaminoakes I added a docker-compose.yml file for local development and refactored a bit around the "copy . ."  Let's me know if it address this one

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+
+require "stringio"
+require_relative "../lib/rails_upgrade_analyzer"
+require_relative "../../deprecation_summary"
```
Should these be in the load path?

benjaminoakes replied:
Also, it feels strange having a `bin` directory within `lib`.  Can we move this part into the toplevel `bin`?

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+    RAILS_4_ACCEPTED = "Rails 4 Accepted"
+    RAILS_4_REJECTED = "Rails 4 Rejected"
+    RAILS_4_REBASE = "Rails 4 CI Needed (rebase master)"
+    RAILS_4_WARNING = "Rails 4 Check Deprecation Warnings"
```
What are your thoughts on making these a little more generic?  E.g. so we don't have to change these for Rails 5.

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,168 @@
+      link = "<a href='https://travis-ci.com/Liaison-Intl/WebAdMIT/jobs/#{job.id})'>#{name}</a>"
+      body = job.log.clean_body
+      match = body.match(/(?<tests>\d+) tests, (?<passed>\d+) passed, (?<failures>\d+) failures, (?<errors>\d+) errors/)
+      summary = DeprecationSummary.new(StringIO.new(body).readlines)
```
`StringIO.new(body).readlines`?  Is that basically doing `body.split("\n")`?

benjaminoakes commented on lib/ci/lib/rails_upgrade_analyzer.rb:
```diff
@@ -0,0 +1,11 @@
+  autoload :AuthenticationError, "rails_upgrade_analyzer/github_proxy"
+  autoload :GithubProxy, "rails_upgrade_analyzer/github_proxy"
+  autoload :JobResult, "rails_upgrade_analyzer/job_result"
+  autoload :ResultReporter, "rails_upgrade_analyzer/result_reporter"
```
If I remember right, Matz isn't a fan of `autoload` (he'd like to remove it), so we should maybe avoid using it in new code...  In this case, we could just do normal calls to `require`.

benjaminoakes commented on lib/ci/lib/rails_upgrade_analyzer/github_proxy.rb:
```diff
@@ -0,0 +1,43 @@
+
+    def remove_label(label)
+      client.remove_label(repo_name, pull_request_number, label)
+    rescue
```
Can we be more specific here?

kstevens715 replied:
Catch a specific class of error.

benjaminoakes commented on lib/ci/lib/rails_upgrade_analyzer/github_proxy.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    def client
+      @client || fail("You must login first")
```
Why not call `login`?

benjaminoakes commented on lib/ci/lib/rails_upgrade_analyzer/job_result.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+
+    def deprecation_count
+      deprecations.each_value.inject(&:+) || 0
```
`deprecations.values.inject(&:+)` seems a little more natural, if I understand the point of this code...

benjaminoakes commented on lib/ci/lib/rails_upgrade_analyzer/result_reporter.rb:
```diff
@@ -0,0 +1,112 @@
+        table.tbody do |tbody|
+          all_results.each do |result|
+            tbody.tr do |tr|
+              tr << "<td>#{result.description}</td>\n"
```
Wasn't possible to use the builder here?

kstevens715 replied:
No, the description itself actually contains HTML, and using builder escapes the HTML.

benjaminoakes commented on lib/deprecation_summary.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def deprecation_count
+    deprecations.each_value.inject(&:+) || 0
```
Same note about `values` here...

benjaminoakes commented on test/github_proxy_test.rb:
```diff
@@ -0,0 +1,55 @@
+    yield
+  ensure
+    ENV[name] = value_was
+  end
```
I generally prefer passing in `ENV` (dependency injection) and then using hashes in the tests.  A little easier to reason about.

This isn't _wrong_, it's just that I've seen this changing `ENV` cause hard-to-track issues before.

benjaminoakes commented on test/result_reporter_test.rb:
```diff
@@ -0,0 +1,110 @@
+    assert_equal "</tr>", line(reporter, 25)
+
+    assert_equal "</tbody>", line(reporter, 26)
+    assert_equal "</table>", line(reporter, 27)
```
This took me a second to realize what was happening.  Almost feels like we should check equality against a docstring or array or something...

Just a bit surprising.  Most times, I wouldn't expect HTML to be checked this way.

**Edit:** Also worth noting that these tests are pretty brittle.  You could change the implementation so valid HTML is still generated (e.g. remove whitespace/add whitespace), but these assertions would fail.

kstevens715 replied:
Maybe use Nokogiri

benjaminoakes commented on lib/ci/lib/rails_upgrade_analyzer.rb:
```diff
@@ -0,0 +1,11 @@
+require "octokit"
+require "travis"
+
+$LOAD_PATH.unshift(File.dirname(__FILE__))
```
`lib` should already be in the load path.  Why are we nesting another `lib` inside it?

We have `lib/ci/lib/`, but we should probably have `lib/ci/`.

benjaminoakes commented on lib/ci/lib/rails_upgrade_analyzer/github_proxy.rb:
```diff
@@ -0,0 +1,43 @@
+
+  class GithubProxy
+    def self.access_token
+      ENV.fetch("GITHUB_TOKEN") do
```
I mentioned this elsewhere, but probably best to pass in a reference to `ENV` rather than hardcoding a dependency on it here...

kstevens715 committed:
[NF-8840] Clear cache before each analysis

Each pull request is compared against the latest master build. However,
the Travis gem caches the enumerable used to get the latest builds and
so once you run it once it will never see newer builds unless you
explicitely clear the cache. This was causing the rails upgrade analyzer
to compare each PR to the first master build that was grabbed when the
process started up. For example, after the process had been up and
running for five consecutive days new PR's were still being compared to
a master build from five days ago.

kstevens715 committed:
[NF-8840] Take a build's base branch into account

kstevens715 commented on lib/ci/lib/rails_upgrade_analyzer/result_reporter.rb:
```diff
@@ -0,0 +1,112 @@
+    end
+
+    def add_deprecation_report_header(table)
+      headers = ["Deprecation", "Result 1", "Result 2", "Difference"]
```
Maybe change "Result 1" and "Result 2" to describe which build/branch it's referring to.

kstevens715 commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,176 @@
+
+      puts "Analyzing PR: #{build.pull_request_number}"
+      results = analyze_build(build, "Current")
+      puts "Getting master results"
```
A couple things here mention, "master". That's no longer accurate.

benjaminoakes commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,176 @@
+    end
+
+    def base_branch(build)
+      match = build.branch_info.match(/\A[^\s]*/)
```
I **think** this method could be...

``` ruby
def base_branch(build)
  build.branch_info.match(/\A[^\s]*/).to_s
end
```

...but this is something I just discovered recently, so I might be wrong about that.  :smile:

kstevens715 commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,176 @@
+        reporter = ResultReporter.new(master_result, result)
+
+        if reporter.deprecation_warnings_changed?
+          github.add_labels_to_an_issue([RAILS_4_WARNING])
```
This doesn't seem to be working.

kstevens715 commented on lib/ci/bin/rails_upgrade_analyzer:
```diff
@@ -0,0 +1,176 @@
+        comment << "</p></details><br />"
+      end
+
+      overall_failing = (any_failing || overall_feature_result.passing_percent < overall_master_result.passing_percent)
```
Might be able to simplify this by just using `any_failing`.

kstevens715 committed:
[NF-8840] Generalize upgrade script

The upgrade script no longer refers to Liaison specific projects. Also,
it does not apply just to Rails 4 or even Rails, for that matter.

As part of generalizing this, I had to remove some hard coded values so
I also introduced OptionParser to pass in command line options.

kstevens715 committed:
[NF-8840] Gemfile.lock Ruby version should match Travis

kstevens715 committed:
[NF-8840] Remove unnecessary login method

Now the class will just login automatically as needed

kstevens715 committed:
[NF-8840] Catch a more specific error

* Catch more specific errors
* Use #values instead of #each_value because it's more natural and more
  developers are aware of it.

pfcarrier committed:
Improve readability of Dockerfile.upgrade_analyzer

Preseve the two hacks end-result that previsouly allowed to:
a) insure that 'apk add' always use latest avail package by
   now using 'apk' --no-cache switch

b) minimize size of all installed gem by not including doc by
   now setting '--no-ri --no-rdoc' as system wide option for all gem
   install in /etc/gemrc.
   Bonus:  this shave ~3 megs from the docker images.  213 -> 209 megs

pfcarrier committed:
update README to reflect new location of GKE cluster

pfcarrier committed:
Fix kubectl deployments command -- allow to pass ARG with success

pfcarrier committed:
* code is now in /opt/ci instead of / in the Docker container
* transition to declarative deployment instead of imperative ( new
  file: kubectl_config/upgrade_analyzer_deployment.yml)
* now using GKE inside 'wa-qa-1087' GCE projects and registry instead of
  Kyle's personal GKE project
* TODO: Use GITHUB_TOKEN via env variable instead of hard-coding it
        in kubectl_config/upgrade_analyzer_deployment.yml
* TODO: Add travisci.yml incantation to fully deploy to GKE at each
        push to 'master' using secret defined as env var inside travisci
        configuration.

W/ Mehar

pfcarrier committed:
user kubernete secret facility to store secret within GKE cluster ( GITHUB_TOKEN )

kstevens715 committed:
[NF-8840] Use better build descriptions in output

kstevens715 committed:
[NF-8840] Reduce test brittleness w/Nokogiri

kstevens715 committed:
[NF-8840] Extract ERB template: Step 1

kstevens715 committed:
[NF-8840] Extract ERB template: Step 2

benjaminoakes commented on Dockerfile.upgrade_analyzer:
```diff
@@ -0,0 +1,31 @@
+    gem install \
+    bundler \
+    rdoc \
+    io-console
```
I said this elsewhere, but why did we drop the versions here?

benjaminoakes replied:
@pfcarrier ^

pfcarrier replied:
good catch.  fixed.

kstevens715 committed:
[NF-8840] Refactor large method

kstevens715 committed:
[NF-8840] Don't set instance variable in view

pfcarrier committed:
scrub company scpecific information/name
introduce best-practice way to deploy to GKE declaratively using
the 'deployment' object stored on the GKE cluster.

pfcarrier committed:
add resources limit and request to pods declaration

per best practice to allow GKE to efficiently binpack various
container on the cluster.

pfcarrier committed:
travis deploy to GKE on push to 'master'

pfcarrier committed:
add support for docker-compose for local development env.

benjaminoakes commented on CD/deploy.sh:
```diff
@@ -0,0 +1,29 @@
+### On any error the script will fail
+set -xeo pipefail
+
+### Install dependancy
```
dependancy -> dependency

benjaminoakes commented on README.md:
```diff
@@ -10,6 +10,82 @@ Utilities to help upgrading a ~250K line Rails app.
   * Rails tests
   * Jasmine specs
 
+## Pull Request Analyzer
```
A lot of the formatting in this file isn't quite what we should have for Markdown.  Can we indent or use code fences (```) around the commands?

benjaminoakes commented on .travis.yml:
```diff
@@ -0,0 +1,24 @@
+  - docker
+
+script:
+  - CD/deploy.sh
```
What does `CD` stand for?  Continuous Delivery/Deployment?  It's a bit more terse than it maybe needs to be...

benjaminoakes commented on README.md:
```diff
@@ -10,6 +10,82 @@ Utilities to help upgrading a ~250K line Rails app.
+kubectl config view
+kubectl proxy
+kubectl cluster-info # see URI for kubernetes-dashboard
+# ^^  See your google cloud console for the credential of the GKE cluster
```
This doesn't seem like it's a heading...

benjaminoakes commented on README.md:
```diff
@@ -20,7 +96,31 @@ Install Docker and look at the Docker instructions.
+docker-compose up
+# ^^ Use the CI container as a runtime with your local directory mounted
+#    inside the container so that the code in this directory is used live.
+#    allow to develop in the exact same env in which the code will run.
```
This doesn't seem like it's a heading...

benjaminoakes commented on bin/helper_upgrade_analyzer:
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
```
Maybe we should call this file `start_upgrade_analyzer`?  "Helper" makes me think of other concepts...

pfcarrier replied:
yeah, that was ugly, a big hack around docker-compose issue using interpolation in 'command:'.
Mehar made the needs for that file obsolete ( upgrade_analyzer can now use ENV var directly when available )

benjaminoakes commented on CD/deploy.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+### On any error the script will fail
+set -xeo pipefail
```
Cool...  How is this different than `set -o errexit`?

pfcarrier replied:
set '-e' is a shortcut for '-o errexit'
'-o pipefail' add protection against error happening in any commands forming a pipeline.
otherwise only the last command return code is checked.  So error in other command would
not lead to the script exiting with an error.
ex:  
echo "a" | sed -e -badswitch | cat
echo $?
0

benjaminoakes commented on CD/deploy.sh:
```diff
@@ -0,0 +1,29 @@
+gcloud --version
+
+### Authenticate to gcloud
+export GOOGLE_APPLICATION_CREDENTIALS=/tmp/gce.json # ENV var used by kubectl
```
> ENV var used by kubectl

Good comment.  👏

benjaminoakes commented on README.md:
```diff
@@ -10,6 +10,82 @@ Utilities to help upgrading a ~250K line Rails app.
+### Deploy declaratively -- first time
+kubectl create -f kubectl_config/upgrade-analyzer-deployment.yml
+kubectl edit deployment/upgrade-analyzer-deployment
+# ^^ replace NAME_OF_REPO with with the repo you want to monitor
```
This doesn't seem like it's a heading...

benjaminoakes commented on README.md:
```diff
@@ -10,6 +10,82 @@ Utilities to help upgrading a ~250K line Rails app.
+### Deploy declaratively -- when already exist
+kubectl set image deployment/upgrade-analyzer-deployment \
+  upgrade-analyzer=gcr.io/<google-project>/ci:<build-number>
+# ^^ deploy will not occur if you don't set a new image build-number
```
This doesn't seem like it's a heading...

benjaminoakes reviewed (changes requested):
Lots of interesting stuff, but some cleanup is needed.

devdude committed:
[NF-8840] Use ENV vars and improve onboarding

Clarify readme.md instructions and validate them.
Using env vars fallback with upgrade-analyzer
scrub one occurence of company name
deprecate helper_upgrade_analyzer
improve readability of TEMPLATE in k8s API file
add version number for rubygem

w/ Pierre.

pfcarrier committed:
[NF-8840] improve clarity on script naming

rename CD/deploy.sh to travis/continuous_deployment

pfcarrier committed:
[NF-8840] help keep docker image tidy, adding ignore

pfcarrier committed:
[NF-8840] Add instruction for quick evaluation of upgrade_analyzer by users

Point to the newly setup automated build on
https://hub.docker.com/r/liaisonintl/upgrade_toolbelt/

jcreed reviewed (approved):
I think for an initial pr this looks fine at this point with the recent revisions.  It's a very good starting point for this project which can only get better.

loto commented on bin/upgrade_analyzer:
```diff
@@ -0,0 +1,211 @@
+require_relative "../lib/upgrade_analyzer"
+
+module UpgradeAnalyzer
+  class Listener
```
Why didn't you extract the class from here?

loto reviewed (approved):
👍

benjaminoakes commented on Dockerfile.upgrade_analyzer:
```diff
@@ -0,0 +1,30 @@
+RUN bundle install
+RUN mkdir -p /opt/ci
+
+COPY . /opt/ci
```
How does this work for development within the container?  Shouldn't we mount the directory instead?

pfcarrier replied:
Its a pattern I see used a lots.  It's allow to use the same container for both development and testing.  For development docker-compose mount the directory into the container.  For testing we use only what is in the container.

benjaminoakes commented on README.md:
```diff
@@ -19,8 +35,148 @@ Install Docker and look at the Docker instructions.
+# ^^ deploy will not occur if you don't set a new image build-number
+#    if you want to force a pods recreation with same ci:<buildnumber>
+#    simply delete the running pods with 'kubectl delete pods POD_NAME'
+#    the replication controller will automatically recreate a new one
```
The formatting of this file is still not what  is needed to be a properly formatted Markdown document.  These level-one headings look like they were probably meant to be code comments. This is our README and is front and center for anyone that stumbles across this project, so we should clean this up.

benjaminoakes commented on README.md:
```diff
@@ -19,8 +35,148 @@ Install Docker and look at the Docker instructions.
+kubectl config use-context default
 
-### Docker
+## Docker for the upgrade_toolbelt tool ( legacy )
```
What are we thinking of doing to resolve this?  We're still using the original scripts I wrote to get summary information after the fact, aren't we?

benjaminoakes commented on .travis.yml:
```diff
@@ -0,0 +1,29 @@
+rvm:
+  - 2.3.1
+services:
+  - docker
```
If we aren't already, we should (eventually) run the tests in the Docker container.

benjaminoakes commented on docker-compose.yml:
```diff
@@ -0,0 +1,13 @@
+    env_file:
+      - ./kv.conf
+    volumes:
+      - .:/opt/ci
```
I played with `docker-compose` as a part of reviewing this.  Definitely should replace my "cheat sheet", but hey, I'm still learning this stuff. 😄

benjaminoakes commented on bin/upgrade_analyzer:
```diff
@@ -0,0 +1,211 @@
+  opts.on("-rREPOSITORY_NAME", "--repo=REPOSITORY_NAME", "GitHub repository to use. Include organization/user name and repo name") do |repo|
+    options[:repo] = repo
+  end
+  opts.on("-t[GITHUB_TOKEN]", "--token[=GITHUB_TOKEN]", "Your personal GitHub token") do |github_token|
```
I know this has already been discussed, but it's probably best to prefer environment variables to command-line options, especially for the token.  (Keep in mind that the host system can see what commands are running in the containers...  It's kind of like providing your password in a command line argument.)

pfcarrier replied:
A bit lower in the file there is a fallback to using ENV var ( if it exist ) when command line switch is not provided.  Indeed for security using the ENV would be preferable.  Though, somehow I love that we support both options, leaving the choice to the user depending of her use case.

benjaminoakes commented on lib/result_reporter.erb:
```diff
@@ -0,0 +1,63 @@
+  </summary>
+  <p>
+
+  <br /><strong>Test report</strong><br />
```
This seems like a strange way of marking up what is essentially a heading...

benjaminoakes commented on lib/result_reporter.erb:
```diff
@@ -0,0 +1,63 @@
+      <% end %>
+    </tbody>
+  </table>
+  </p>
```
Using a `<p>` for this seems forced...

benjaminoakes commented on lib/result_reporter.erb:
```diff
@@ -0,0 +1,63 @@
+    </tbody>
+  </table>
+
+  <strong>Deprecation report</strong><br /><br />
```
This seems like a strange way of marking up what is essentially a heading...

benjaminoakes commented on lib/result_reporter.rb:
```diff
@@ -0,0 +1,57 @@
+    end
+
+    def report
+      RESULT_REPORTER_TEMPLATE.result(binding).gsub("\n", "")
```
I like how this refactoring went.  🌟

benjaminoakes commented on lib/upgrade_analyzer.rb:
```diff
@@ -0,0 +1,11 @@
+$LOAD_PATH.unshift(File.dirname(__FILE__))
```
Isn't `lib` normally in the load path?  I normally consider changing `$LOAD_PATH` to be a sign of something that shouldn't be done that way...

benjaminoakes reviewed (changes requested):
Getting close...  The only thing that I really want addressed before merging is the formatting in the README.

pfcarrier committed:
Address some of the README.md format issue

pfcarrier committed:
[NF-8840] run tests from within the container

Test run from within the build container
move container build from deploy to before_script
rework logic that control which branch is deployed

pfcarrier committed:
[NF-8840] run apps as user 'ci' instead or 'root'

run apps as non root to follow recommended best-practice
remove non-needed rdoc gem

benjaminoakes commented on README.md:
```diff
@@ -19,9 +35,181 @@ Install Docker and look at the Docker instructions.
+### Push your local container to GCE so it can be consumed by Kubernetes
+  ``
+  gcloud docker push gcr.io/<google-project>/ci:<build-number>
+  ``
```
This should have three backticks, not two.

pfcarrier replied:
fixed

benjaminoakes commented on README.md:
````diff
@@ -19,9 +35,181 @@ Install Docker and look at the Docker instructions.
+  kubectl config use-context default
+  ```
+## Docker for the upgrade_toolbelt tool ( legacy )
+  ``
````
This should have three backticks, not two.

pfcarrier replied:
ended up pasting the whole file in http://dillinger.io/
hopefully didn't miss anything else.

benjaminoakes reviewed (changes requested):
Getting closer on the README.  Reminder: you can click the "View" button and you'll see it rendered, which makes the errors easier to spot.

pfcarrier committed:
[NF-8840] add instruction to access 'Logging' facility of GKE

tidy up the README.md using http://dillinger.io/

devdude reviewed (approved):
Let's get this merged as it stands and we can loop back on the outstanding issues later.

benjaminoakes commented on .travis.yml:
```diff
@@ -0,0 +1,28 @@
+  - docker build -t gcr.io/${GCE_PROJECT}/ci:${TRAVIS_BUILD_NUMBER} -f Dockerfile.upgrade_analyzer .
+
+script:
+  - docker run gcr.io/${GCE_PROJECT}/ci:${TRAVIS_BUILD_NUMBER} bundle exec rake
```
Nice!

benjaminoakes reviewed (approved):
Markdown looks good now!  Merging...
